thermohl-docs/docs/api-reference/parameters.md
<!---
SPDX-FileCopyrightText: 2025 RTE (https://www.rte-france.com)

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at http://mozilla.org/MPL/2.0/.
SPDX-License-Identifier: MPL-2.0
--->

# Parameters and Default Values

Solvers in ThermOHL take a dictionary as an argument, where all keys
are strings and all values are either integers, floats or 1D
`numpy.ndarray` of integers or floats. It is important to note that
all arrays should have the same size.  Missing or `None` values in the
input dictionary are replaced with a default value, available using
`solver.default_values()`.

Below is a table with all physical parameters used in ThermOHL, with
units, default values and in which set of power terms they are used.

[TODO : check with sources + add a paragraph explaining which parameters are not required if we use solar_irradiance directly, or measured_global_radiation for the RTE model]: #

| Parameter                   | Default Value | Unit            | Used in CIGRE | Used in IEEE | Used in OLLA | Used in RTE | Comment                                                                                                                                                                                                                                                                   |
|-----------------------------|---------------|-----------------|---------------|--------------|--------------|-------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| latitude                    | 45            | degree          | yes           | yes          | yes          | yes         | latitude                                                                                                                                                                                                                                                                  |
| longitude                   | 0             | degree          | no            | no           | no           | no          | longitude                                                                                                                                                                                                                                                                 |
| altitude                    | 0             | linear_mass     | yes           | yes          | yes          | yes         | altitude                                                                                                                                                                                                                                                                  |
| cable_azimuth               | 0             | degree          | yes           | yes          | yes          | yes         | cable_azimuth                                                                                                                                                                                                                                                             |
| datetime_utc                | "2000-03-21T12:00:00"             | N/A             | yes           | yes          | yes          | yes         | date (numpy.datetime64 formated "yyyy-mm-ddThh:MM:ss")                                                                                                                                                                                                                    |
| ambient_temperature         | 15            | celsius         | yes           | yes          | yes          | yes         | ambient temperature                                                                                                                                                                                                                                                       |
| wind_speed                  | 0             | linear_mass.s⁻¹ | yes           | yes          | yes          | yes         | wind speed                                                                                                                                                                                                                                                                |
|  wind_azimuth        | 90            | degree     | yes           | yes          | yes          | yes         | wind azimuth (regarding north)                                 |
| wind_attack_angle        | None            | radian     | yes           | yes          | yes          | yes         | (absolute) angle between the wind and the cable. Must be comprised between 0 and π / 2. If provided, this overrides the angle computed from wind_azimuth and cable_azimuth.                                  |
| albedo                      | 0.8           | N/A             | yes           | no           | no           | no          | albedo                                                                                                                                                                                                                                                                    |
| turbidity                   | 0.1           | N/A             | no            | yes          | no           | no          | coefficient for air pollution from 0 (clean) to 1 (polluted)                                                                                                                                                                                                              |
| transit                     | 100           | A               | yes           | yes          | yes          | yes         | transit intensity                                                                                                                                                                                                                                                         |
| linear_mass                 | 1.5           | kg.m⁻¹          | yes           | yes          | yes          | yes         | mass per unit length (only used in transient mode)                                                                                                                                                                                                                        |
| core_diameter               | 1.9E-02       | linear_mass     | no            | no           | yes          | yes         | core diameter                                                                                                                                                                                                                                                             |
| outer_diameter              | 3.0E-02       | linear_mass     | yes           | yes          | yes          | yes         | external (global) diameter                                                                                                                                                                                                                                                |
| core_area                   | 2.84E-04      | m²              | no            | no           | yes          | yes         | core section                                                                                                                                                                                                                                                              |
| A                           | 7.07E-04      | m²              | no            | no           | yes          | yes         | external (global) section                                                                                                                                                                                                                                                 |
| roughness_ratio             | 4.0E-02       | N/A             | yes           | no           | no           | no          | roughness                                                                                                                                                                                                                                                                 |
| radial_thermal_conductivity | 1.0           | W.m⁻¹.K⁻¹       | no            | no           | yes          | yes         | radial thermal conductivity (not used in 1t equation)                                                                                                                                                                                                                     |
| heat_capacity               | 500           | J.kg⁻¹.K⁻¹      | yes           | yes          | yes          | yes         | specific heat capacity (only used in transient mode)                                                                                                                                                                                                                      |
| solar_absorptivity          | 0.5           | N/A             | yes           | yes          | yes          | yes         | solar absorption                                                                                                                                                                                                                                                          |
| emissivity                  | 0.5           | N/A             | yes           | yes          | yes          | yes         | emissivity                                                                                                                                                                                                                                                                |
| linear_resistance_dc_20c    | 2.5E-05       | Ohm.m⁻¹         | yes           | no           | yes          | yes         | electric resistance per unit length (DC) at 20°C                                                                                                                                                                                                                          |
| magnetic_coeff              | 1.006         | N/A             | yes           | no           | yes          | yes         | coefficient for magnetic effects                                                                                                                                                                                                                                          |
| magnetic_coeff_per_a        | 0.016         | A⁻¹             | no            | no           | yes          | yes         | coefficient for magnetic effects                                                                                                                                                                                                                                          |
| temperature_coeff_linear    | 3.8E-03       | K⁻¹             | yes           | no           | yes          | yes         | linear resistance augmentation with temperature                                                                                                                                                                                                                           |
| temperature_coeff_quadratic | 8.0E-07       | K⁻²             | no            | no           | yes          | yes         | quadratic resistance augmentation with temperature                                                                                                                                                                                                                        |
| linear_resistance_temp_high | 3.05E-05      | Ohm.m⁻¹         | no            | yes          | no           | no          | electric resistance per unit length (DC) at temp_high                                                                                                                                                                                                                     |
| linear_resistance_temp_low  | 2.66E-05      | Ohm.m⁻¹         | no            | yes          | no           | no          | electric resistance per unit length (DC) at temp_low                                                                                                                                                                                                                      |
| temp_high                   | 60            | celsius         | no            | yes          | no           | no          | temperature for linear_resistance_temp_high measurement                                                                                                                                                                                                                   |
| temp_low                    | 20            | celsius         | no            | yes          | no           | no          | temperature for linear_resistance_temp_high measurement                                                                                                                                                                                                                   |
| measured_global_radiation   | NaN           | W.m⁻²           | no            | no           | no           | yes         | global radiation (not to be confused with solar irradiance. In Rte context solar irradiance is computed from global radiation and other parameters. Global radiation has no meaning outside Rte context.) NB : in our context "irradiance" and "radiation" mean the same. |
| solar_irradiance            | NaN   | W.m⁻²           | yes           | yes          | yes          | no          | NB : in our context "irradiance" and "radiation" mean the same.                                                                                                                                                                                                           |

For consistent joule heating outputs between CIGRE and IEEE joule
power terms, you must have

* $ R_{\text{DC,low}}=R_{\text{DC},20} $;
* $ T_{\text{low}}=20 $;
* any $ T_{\text{high}} > T_{\text{low}} $ and
$ R_{\text{DC,high}} - R_{\text{DC,low}} = (T_{\text{high}} - T_{\text{low}}) \cdot k_{\ell} \cdot R_{\text{DC},20} $.

If you use direct solar radiation formula (with `measured_global_radiation` key), you can
ignore the following parameters : `latitude`, `longitude`, `datetime_utc`, `albedo` and `turbidity`.



src/thermohl/solver/parameters.py
# SPDX-FileCopyrightText: 2025 RTE (https://www.rte-france.com)
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
# SPDX-License-Identifier: MPL-2.0

from typing import Optional, Any, KeysView, Iterable
import numpy as np


class DEFAULT_PARAMETERS:
    tmin = -99.0
    tmax = +999.0
    tol = 1.0e-09
    maxiter = 64
    imin = 0.0
    imax = 9999.0


class Parameters:
    """Object to store Solver args in a dict-like manner."""

    def __init__(self, input_dict: Optional[dict[str, Any]] = None):
        # add default values
        self._set_default_values()
        # use values from input dict
        if input_dict is None:
            return
        arg_keys = self.keys()
        for input_key, input_value in input_dict.items():
            if input_key in arg_keys and input_value is not None:
                self[input_key] = input_value

    def _set_default_values(self) -> None:
        """Set default values."""
        # position
        self.latitude = 45.0  # latitude (deg)
        self.longitude = 0.0  # longitude (deg)
        self.altitude = 0.0  # altitude (m)
        self.cable_azimuth = 0.0  # cable_azimuth (deg)
        self.datetime_utc = np.datetime64("2000-03-21T12:00:00")

        # weather and mesurement
        self.measured_global_radiation = np.nan
        self.solar_irradiance = np.nan
        self.ambient_temperature = 15.0  # ambient temperature (C)
        self.ambient_pressure = 1.0e05  # ambient pressure (Pa)
        self.relative_humidity = 0.8  # relative humidity (none, in [0, 1])
        self.precipitation_rate = 0.0  # rain precipitation rate (m.s**-1)
        self.wind_speed = 0.0  # wind speed (m.s**-1)
        self.wind_azimuth = 90.0  # wind_azimuth (deg, regarding north)
        self.wind_attack_angle = np.nan  # wind attack angle (rad)
        self.nebulosity = np.nan  # nebulosity (1)
        self.albedo = 0.15  # albedo (1)
        # coefficient for air pollution from 0 (clean) to 1 (polluted)
        self.turbidity = 0.1
        self.transit = 100.0  # transit intensity (A)

        # conductor
        self.linear_mass = 1.5  # mass per unit length (kg.m**-1)
        self.core_diameter = 1.9e-02  # core diameter (m)
        self.outer_diameter = 3.0e-02  # external (global) diameter (m)
        self.core_area = 2.84e-04  # core section (m**2)
        self.outer_area = 7.07e-04  # external (global) section (m**2)
        self.roughness_ratio = 4.0e-02  # roughness (1)
        # radial thermal conductivity (W.m**-1.K**-1)
        self.radial_thermal_conductivity = 1.0
        self.heat_capacity = 500.0  # specific heat capacity (J.kg**-1.K**-1)
        self.solar_absorptivity = 0.5  # solar absorption (1)
        self.emissivity = 0.5  # emissivity (1)
        # electric resistance per unit length (DC) at 20°C (Ohm.m**-1)
        self.linear_resistance_dc_20c = 2.5e-05
        self.magnetic_coeff = 1.006  # coefficient for magnetic effects (1)
        self.magnetic_coeff_per_a = 0.016  # coefficient for magnetic effects (A**-1)
        # linear resistance augmentation with temperature (K**-1)
        self.temperature_coeff_linear = 3.8e-03
        # quadratic resistance augmentation with temperature (K**-2)
        self.temperature_coeff_quadratic = 8.0e-07
        # electric resistance per unit length (DC) at temp_high (Ohm.m**-1)
        self.linear_resistance_temp_high = 3.05e-05
        # electric resistance per unit length (DC) at temp_low (Ohm.m**-1)
        self.linear_resistance_temp_low = 2.66e-05
        # temperature for linear_resistance_temp_high measurement (°C)
        self.temp_high = 60.0
        # temperature for linear_resistance_temp_low measurement (°C)
        self.temp_low = 20.0

    def keys(self) -> KeysView[str]:
        """Get list of members as dict keys."""
        return self.__dict__.keys()

    def __getitem__(self, key: str) -> Any:
        return self.__dict__[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.__dict__[key] = value

    def get_number_of_computations(self) -> int:
        return max(
            (len(self[k]) for k in self.keys() if isinstance(self[k], Iterable)),
            default=1,
        )

    def extend(self) -> None:
        """
        Extend all compressed elements in the Args dictionary to the right length, ie the number of computations.
        If the element is a list, it already has the right length.
        If the element is a scalar, it is replaced with a list of the right length filled with the scalar value.
        """
        number_of_computations = self.get_number_of_computations()
        for key in self.keys():
            if not isinstance(self[key], Iterable):
                self[key] = np.array(number_of_computations * [self[key]])

    def compress(self) -> None:
        """
        Compress the elements in the Args dictionary by replacing lists containing a unique value
        with this value.
        """
        for key in self.keys():
            u = np.unique(self[key])
            if len(u) == 1:
                self[key] = u[0]
